File: Python/OneLang/One/Compiler.py
from OneLangStdLib import *
import OneLang.One.Ast.Types as types
import OneLang.Parsers.TypeScriptParser2 as typeScrPars
import OneLang.StdLib.PackageManager as packMan
import OneLang.StdLib.PackagesFolderSource as packFoldSourc
import OneLang.One.Transforms.FillParent as fillPar
import OneLang.One.Transforms.FillAttributesFromTrivia as fillAttrFromTriv
import OneLang.One.Transforms.ResolveGenericTypeIdentifiers as resGenTypeIdents
import OneLang.One.Transforms.ResolveUnresolvedTypes as resUnrTypes
import OneLang.One.Transforms.ResolveImports as resImps
import OneLang.One.Transforms.ConvertToMethodCall as convToMethCall
import OneLang.One.Transforms.ResolveIdentifiers as resIdents
import OneLang.One.Transforms.InstanceOfImplicitCast as instOfImplCast
import OneLang.One.Transforms.DetectMethodCalls as detMethCalls
import OneLang.One.Transforms.InferTypes as inferTypes
import OneLang.One.Transforms.CollectInheritanceInfo as collInhInfo
import OneLang.One.Transforms.FillMutabilityInfo as fillMutInfo
import OneLang.One.AstTransformer as astTrans
import OneLang.One.ITransformer as iTrans
import OneLang.One.Transforms.LambdaCaptureCollector as lambdCaptColl

class Compiler:
    def __init__(self):
        self.pac_man = None
        self.workspace = None
        self.native_file = None
        self.native_exports = None
        self.project_pkg = None
        self.hooks = None
    
    def init(self, packages_dir):
        self.pac_man = packMan.PackageManager(packFoldSourc.PackagesFolderSource(packages_dir))
        self.pac_man.load_all_cached()
    
    def get_transformers(self, for_declaration_file):
        transforms = []
        if for_declaration_file:
            transforms.append(fillPar.FillParent())
            transforms.append(fillAttrFromTriv.FillAttributesFromTrivia())
            transforms.append(resImps.ResolveImports(self.workspace))
            transforms.append(resGenTypeIdents.ResolveGenericTypeIdentifiers())
            transforms.append(resUnrTypes.ResolveUnresolvedTypes())
            transforms.append(fillMutInfo.FillMutabilityInfo())
        else:
            transforms.append(fillPar.FillParent())
            transforms.append(fillAttrFromTriv.FillAttributesFromTrivia())
            transforms.append(resImps.ResolveImports(self.workspace))
            transforms.append(resGenTypeIdents.ResolveGenericTypeIdentifiers())
            transforms.append(convToMethCall.ConvertToMethodCall())
            transforms.append(resUnrTypes.ResolveUnresolvedTypes())
            transforms.append(resIdents.ResolveIdentifiers())
            transforms.append(instOfImplCast.InstanceOfImplicitCast())
            transforms.append(detMethCalls.DetectMethodCalls())
            transforms.append(inferTypes.InferTypes())
            transforms.append(collInhInfo.CollectInheritanceInfo())
            transforms.append(fillMutInfo.FillMutabilityInfo())
            transforms.append(lambdCaptColl.LambdaCaptureCollector())
        return transforms
    
    def setup_native_resolver(self, content):
        self.native_file = typeScrPars.TypeScriptParser2.parse_file(content)
        self.native_exports = types.Package.collect_exports_from_file(self.native_file, True)
        for trans in self.get_transformers(True):
            trans.visit_files([self.native_file])
    
    def new_workspace(self, pkg_name = "@"):
        self.workspace = types.Workspace()
        for intf_pkg in self.pac_man.interfaces_pkgs:
            lib_name = f'''{intf_pkg.interface_yaml.vendor}.{intf_pkg.interface_yaml.name}-v{intf_pkg.interface_yaml.version}'''
            self.add_interface_package(lib_name, intf_pkg.definition)
        
        self.project_pkg = types.Package(pkg_name, False)
        self.workspace.add_package(self.project_pkg)
    
    def add_interface_package(self, lib_name, definition_file_content):
        lib_pkg = types.Package(lib_name, True)
        file = typeScrPars.TypeScriptParser2.parse_file(definition_file_content, types.SourcePath(lib_pkg, types.Package.index))
        self.setup_file(file)
        lib_pkg.add_file(file, True)
        self.workspace.add_package(lib_pkg)
    
    def setup_file(self, file):
        file.add_available_symbols(self.native_exports.get_all_exports())
        file.literal_types = types.LiteralTypes((file.available_symbols.get("TsBoolean")).type, (file.available_symbols.get("TsNumber")).type, (file.available_symbols.get("TsString")).type, (file.available_symbols.get("RegExp")).type, (file.available_symbols.get("TsArray")).type, (file.available_symbols.get("TsMap")).type, (file.available_symbols.get("Error")).type, (file.available_symbols.get("Promise")).type)
        file.array_types = [(file.available_symbols.get("TsArray")).type, (file.available_symbols.get("IterableIterator")).type, (file.available_symbols.get("RegExpExecArray")).type, (file.available_symbols.get("TsString")).type, (file.available_symbols.get("Set")).type]
    
    def add_project_file(self, fn, content):
        file = typeScrPars.TypeScriptParser2.parse_file(content, types.SourcePath(self.project_pkg, fn))
        self.setup_file(file)
        self.project_pkg.add_file(file)
    
    def process_files(self, files):
        for trans in self.get_transformers(False):
            trans.visit_files(files)
            if self.hooks != None:
                self.hooks.after_stage(trans.name)
    
    def process_workspace(self):
        for pkg in list(filter(lambda x: x.definition_only, self.workspace.packages.values())):
            for trans in self.get_transformers(True):
                trans.visit_files(pkg.files.values())
        
        self.process_files(self.project_pkg.files.values())

File: PHP/src/OneLang/One/Compiler.php
<?php

namespace OneLang\One\Compiler;

use OneLang\One\Ast\Types\Workspace;
use OneLang\One\Ast\Types\Package;
use OneLang\One\Ast\Types\SourcePath;
use OneLang\One\Ast\Types\SourceFile;
use OneLang\One\Ast\Types\ExportedScope;
use OneLang\One\Ast\Types\LiteralTypes;
use OneLang\One\Ast\Types\Class_;
use OneLang\One\Ast\Types\ExportScopeRef;
use OneLang\Parsers\TypeScriptParser\TypeScriptParser2;
use OneLang\StdLib\PackageManager\PackageManager;
use OneLang\StdLib\PackagesFolderSource\PackagesFolderSource;
use OneLang\One\Transforms\FillParent\FillParent;
use OneLang\One\Transforms\FillAttributesFromTrivia\FillAttributesFromTrivia;
use OneLang\One\Transforms\ResolveGenericTypeIdentifiers\ResolveGenericTypeIdentifiers;
use OneLang\One\Transforms\ResolveUnresolvedTypes\ResolveUnresolvedTypes;
use OneLang\One\Transforms\ResolveImports\ResolveImports;
use OneLang\One\Transforms\ConvertToMethodCall\ConvertToMethodCall;
use OneLang\One\Transforms\ResolveIdentifiers\ResolveIdentifiers;
use OneLang\One\Transforms\InstanceOfImplicitCast\InstanceOfImplicitCast;
use OneLang\One\Transforms\DetectMethodCalls\DetectMethodCalls;
use OneLang\One\Transforms\InferTypes\InferTypes;
use OneLang\One\Transforms\CollectInheritanceInfo\CollectInheritanceInfo;
use OneLang\One\Transforms\FillMutabilityInfo\FillMutabilityInfo;
use OneLang\One\AstTransformer\AstTransformer;
use OneLang\One\ITransformer\ITransformer;
use OneLang\One\Transforms\LambdaCaptureCollector\LambdaCaptureCollector;

interface ICompilerHooks {
    function afterStage($stageName);
}

class Compiler {
    public $pacMan;
    public $workspace;
    public $nativeFile;
    public $nativeExports;
    public $projectPkg;
    public $hooks;
    
    function __construct()
    {
        $this->pacMan = null;
        $this->workspace = null;
        $this->nativeFile = null;
        $this->nativeExports = null;
        $this->projectPkg = null;
        $this->hooks = null;
    }
    
    function init($packagesDir) {
        $this->pacMan = new PackageManager(new PackagesFolderSource($packagesDir));
        $this->pacMan->loadAllCached();
    }
    
    function getTransformers($forDeclarationFile) {
        $transforms = array();
        if ($forDeclarationFile) {
            $transforms[] = new FillParent();
            $transforms[] = new FillAttributesFromTrivia();
            $transforms[] = new ResolveImports($this->workspace);
            $transforms[] = new ResolveGenericTypeIdentifiers();
            $transforms[] = new ResolveUnresolvedTypes();
            $transforms[] = new FillMutabilityInfo();
        }
        else {
            $transforms[] = new FillParent();
            $transforms[] = new FillAttributesFromTrivia();
            $transforms[] = new ResolveImports($this->workspace);
            $transforms[] = new ResolveGenericTypeIdentifiers();
            $transforms[] = new ConvertToMethodCall();
            $transforms[] = new ResolveUnresolvedTypes();
            $transforms[] = new ResolveIdentifiers();
            $transforms[] = new InstanceOfImplicitCast();
            $transforms[] = new DetectMethodCalls();
            $transforms[] = new InferTypes();
            $transforms[] = new CollectInheritanceInfo();
            $transforms[] = new FillMutabilityInfo();
            $transforms[] = new LambdaCaptureCollector();
        }
        return $transforms;
    }
    
    function setupNativeResolver($content) {
        $this->nativeFile = TypeScriptParser2::parseFile($content);
        $this->nativeExports = Package::collectExportsFromFile($this->nativeFile, true);
        foreach ($this->getTransformers(true) as $trans)
            $trans->visitFiles(array($this->nativeFile));
    }
    
    function newWorkspace($pkgName = "@") {
        $this->workspace = new Workspace();
        foreach ($this->pacMan->interfacesPkgs as $intfPkg) {
            $libName = $intfPkg->interfaceYaml->vendor . "." . $intfPkg->interfaceYaml->name . "-v" . $intfPkg->interfaceYaml->version;
            $this->addInterfacePackage($libName, $intfPkg->definition);
        }
        
        $this->projectPkg = new Package($pkgName, false);
        $this->workspace->addPackage($this->projectPkg);
    }
    
    function addInterfacePackage($libName, $definitionFileContent) {
        $libPkg = new Package($libName, true);
        $file = TypeScriptParser2::parseFile($definitionFileContent, new SourcePath($libPkg, Package::$INDEX));
        $this->setupFile($file);
        $libPkg->addFile($file, true);
        $this->workspace->addPackage($libPkg);
    }
    
    function setupFile($file) {
        $file->addAvailableSymbols($this->nativeExports->getAllExports());
        $file->literalTypes = new LiteralTypes(($file->availableSymbols->get("TsBoolean"))->type, ($file->availableSymbols->get("TsNumber"))->type, ($file->availableSymbols->get("TsString"))->type, ($file->availableSymbols->get("RegExp"))->type, ($file->availableSymbols->get("TsArray"))->type, ($file->availableSymbols->get("TsMap"))->type, ($file->availableSymbols->get("Error"))->type, ($file->availableSymbols->get("Promise"))->type);
        $file->arrayTypes = array(($file->availableSymbols->get("TsArray"))->type, ($file->availableSymbols->get("IterableIterator"))->type, ($file->availableSymbols->get("RegExpExecArray"))->type, ($file->availableSymbols->get("TsString"))->type, ($file->availableSymbols->get("Set"))->type);
    }
    
    function addProjectFile($fn, $content) {
        $file = TypeScriptParser2::parseFile($content, new SourcePath($this->projectPkg, preg_replace("/\\.ts$/", "", $fn)));
        $this->setupFile($file);
        $this->projectPkg->addFile($file);
    }
    
    function processFiles($files) {
        foreach ($this->getTransformers(false) as $trans) {
            $trans->visitFiles($files);
            if ($this->hooks !== null)
                $this->hooks->afterStage($trans->name);
        }
    }
    
    function processWorkspace() {
        foreach (array_values(array_filter(array_values($this->workspace->packages), function ($x) { return $x->definitionOnly; })) as $pkg)
            foreach ($this->getTransformers(true) as $trans)
                $trans->visitFiles(array_values($pkg->files));
        
        $this->processFiles(array_values($this->projectPkg->files));
    }
}
